VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,863 @@
+  subdirectories `<dut>/dv/env/, <dut>/dv/tests/, <dut>/dv/tb/`.
+* The `<dut>_env_pkg` should import the `<dut>_agent_pkg`
+* All UVM virtual sequences for a given DUT will be placed under `env/seq_lib/`,
+  along with a `<dut>_vseq_list.sv` file that `` `include``-es all of the virtual
+  sequence files. This virtual sequence list file must be `` `include``-ed by
+  `env/<dut>_env_pkg.sv`.
+* If applicable, `env/<dut>_env_pkg.sv` must import the `<dut>_agent_pkg`
+* `tests/<dut>_test_pkg.sv` must import the `<dut>_env_pkg` package
+* All files `` `include``-ed into a single package must be in the same single directory
+
+```systemverilog
+<dut>/dv/
+  env/
+    <dut>_env.sv
+    <dut>_env_pkg.sv
+    <dut>_env_cfg.sv
+    <dut>_env_cov.sv
+    <dut>_scoreboard.sv
+    <dut>_virtual_sequencer.sv
+    seq_lib/
+      // These allow basic functionality to be added to the environment
+      <dut>_base_vseq.sv
+      <dut>_sanity_vseq.sv
+  tb/
+    tb.sv
+  tests/
+    <dut>_base_test.sv
+    <dut>_test_pkg.sv
+```
+* `<block>_agent_pkg.sv, <dut>_env_pkg.sv, <dut>_test_pkg.sv, and tb/tb.sv`
+  must include these `` `include``-es and imports:
+
+```systemverilog
+`include "dv_macros.svh"
+`include "uvm_macros.svh"
+import uvm_pkg::*;
+```
+
+An example environment level package file for an arbitrary block `newblock` should look similar to this:
+
+```systemverilog
+package newblock_env_pkg;
+  // dep packages
+  import uvm_pkg::*;
+  import newblock_agent_pkg::*;
+
+  // macro includes
+  `include "uvm_macros.svh"
+  `include "dv_macros.svh"
+
+  // define any parameters
+  parameter int MY_BLOCK_PARAMETER = 10;
+
+  // define any local types/functions/tasks
+  typedef enum bit {
+    MyblockBooleanZero,
+    MyblockBooleanTrue
+  } myblock_boolean_e;
+
+  // Functions declared within packages must be automatic
+  function automatic bit myblock_test_function(myblock_boolean_e test_boolean);
+    //function definition
+  endfunction
+endpackage
+```
+
+
+## UVM Guidelines
+
+Always call `run_test()` without the test name argument. This allows command
+line test overrides to be performed using `+UVM_TESTNAME=<testname>` without
+having to recompile and allows testlists to be created for regression runs.
+
+
+### Class Definitions
+
+1.  After variable declarations, register any objects and components with the
+    factory using the `uvm_component_utils` and `uvm_object_utils` macros.
+    *   Declare any member variables to be registered with the UVM factory field
+    automation macros `uvm_field_int, uvm_field_enum, uvm_field_object,
+    uvm_field_array_int`, etc. within `uvm_object/component_utils_begin` and
+    `uvm_object/component_utils_end`.
+2.  For all field automation macros, use `UVM_DEFAULT` as the flag argument.
+    For example, if you want to enable everything except comparing and printing for
+    a `rx_delay` field, the macro would look like:
+
+```systemverilog
+`uvm_field_int(rx_delay, UVM_DEFAULT | UVM_NOCOMPARE | UVM_NOPRINT)
```

Comment:
probably want a reminder early here or somewhere that `uvm_<name> is a macro that doesn't require a trailing semicolon

Reply:
Good idea, I included a brief note about this in the UVM library macros section.